protos/sigstore_trustroot.proto

```diff
@@ -162,24 +164,96 @@ message SigningConfig {
         // `https://example.com/ca/api/v2/signingCert`.
         string ca_url = 1;
 
+        // Deprecated: Use openid_connect_provider_urls
         // A URL to an OpenID Connect identity provider.
         //
         // This URL **MUST** be the "base" URL for the OIDC IdP, which clients
         // should perform well-known OpenID Connect discovery against.
         string oidc_url = 2;
 
-        // One or more URLs to Rekor-compatible transparency log.
+        // Deprecated: Use rekor_log_urls
+        // One or more URLs to Rekor-compatible transparency logs.
         //
         // Each URL **MUST** be the "base" URL for the transparency log,
         // which clients should construct appropriate API endpoints on top of.
         repeated string tlog_urls = 3;
 
-        // One ore more URLs to RFC 3161 Time Stamping Authority (TSA).
+        // Deprecated: Use timestamp_authority_urls
+        // One or more URLs to RFC 3161 Time Stamping Authorities (TSA).
         //
         // Each URL **MUST** be the **full** URL for the TSA, meaning that it
         // should be suitable for submitting Time Stamp Requests (TSRs) to
         // via HTTP, per RFC 3161.
         repeated string tsa_urls = 4;
+
+        // URLs to Fulcio-compatible CAs, capable of receiving
+        // Certificate Signing Requests (CSRs) and responding with
+        // issued certificates.
+        //
+        // These URLs **MUST** be the "base" URL for the CAs, which clients
+        // should construct an appropriate CSR endpoint on top of.
+        // For example, if a CA URL is `https://example.com/ca`, then
+        // the client **MAY** construct the CSR endpoint as
+        // `https://example.com/ca/api/v2/signingCert`.
+        //
+        // Clients must select only Service with the highest API version
+        // that the client is compatible with and that is within its
+        // validity period. Clients should select the first Service
+        // that meets this requirement.
+        repeated Service certificate_authority_urls = 6;
+
+        // URLs to OpenID Connect identity providers.
+        //
+        // These URLs **MUST** be the "base" URLs for the OIDC IdPs, which clients
+        // should perform well-known OpenID Connect discovery against.
+        //
+        // Clients must select only Service with the highest API version
+        // that the client is compatible with and that is within its
+        // validity period. Clients should select the first Service
+        // that meets this requirement.
```

Comment:
If the sort order is the same as for TrustedRoot https://github.com/sigstore/protobuf-specs/blob/e943ce11bfbc53bd39035d2bc092a4de3c882c66/protos/sigstore_trustroot.proto#L108-L109 then this could result in a client choosing a service that is valid but about to expire, when a later service in the list would be valid for longer.

Reply:
Good call out. I've added the following:

```
     // All listed Services SHOULD be sorted by the `valid_for` window in
     // descending order, with the newest instance first.
```

With the newest instance first, this should limit how far into the list clients need to search to find a valid instance. Typically a client will just select the first from the list. With a sharding, a client will select the second (old shard) until the first in the list (new shard) becomes valid. For API changes, a client may have to search further into the list, but major API changes will be infrequent.